src/wallet/wallet.cpp

```diff
@@ -2880,21 +2880,31 @@ CBitcoinAddress CWallet::ParseIntoAddress(const CTxDestination& dest, const std:
     return CBitcoinAddress(dest, addrType);
 }
 
+/**
+ * TODO: If we ever add new functionality that modifies the purpose of an address book entry
+ * outside of the below situations, we need to also ensure those situations are considered here
+ */
+bool CWallet::UpdatablePurpose(const std::string& strPurpose1, const std::string& strPurpose2)
+{
+    return (strPurpose1 == AddressBook::AddressBookPurpose::DELEGABLE && strPurpose2 == AddressBook::AddressBookPurpose::DELEGATOR) ||
+           (strPurpose1 == AddressBook::AddressBookPurpose::DELEGATOR && strPurpose2 == AddressBook::AddressBookPurpose::DELEGABLE);
+}
+
 bool CWallet::SetAddressBook(const CTxDestination& address, const std::string& strName, const std::string& strPurpose)
 {
     bool fUpdated = HasAddressBook(address);
+    const bool fUpdatePurpose = !fUpdated || UpdatablePurpose(mapAddressBook.at(address).purpose, strPurpose);
     {
         LOCK(cs_wallet); // mapAddressBook
         mapAddressBook[address].name = strName;
-        if (!strPurpose.empty()) /* update purpose only if requested */
-            mapAddressBook[address].purpose = strPurpose;
+        if (!strPurpose.empty() || fUpdatePurpose) mapAddressBook[address].purpose = strPurpose;
```

Comment:
should probably use `mapAddressBook.at` here too.

I would also remove `!strPurpose.empty()` from the `if` condition here, as we want to update the purpose **only** when `fUpdatePurpose` is true.

Reply:
removed the conditional, but leaving the map access via `[]` operator for the reason stated above.